### PR TITLE
Refactor the core/blocks store as an atomic store

### PR DESCRIPTION
**Pull request by youknowriad**
Related #27109 

This PR refactors the `core/blocks` store as an atomic store. I don't expect performance impact here as it's not a central store that changes a lot but you never know.

It's more of a PR to discover what it takes to refactor a store and  what APIs we might be missing.

### CODE REVIEW

File: docs/designers-developers/developers/data/data-core-blocks.md [hunk 1]
```diff
@@ -12,7 +12,6 @@ Returns block styles by block name.
 
 _Parameters_
 
--   _state_ `Object`: Data state.
```
Comment (youknowriad) [inline docs/designers-developers/developers/data/data-core-blocks.md:15]: I like that it forces us to remove these useless arguments from the docs.

File: packages/blocks/src/store/actions.js [hunk 2]
```diff
@@ -1,191 +1,253 @@
...
+			return uniqBy(
+				[
+					...( blockType.variations ?? [] ),
+					...( currentBlockVariations[ blockType.name ] ?? [] ),
+				],
+				( variation ) => variation.name
+			);
+		} ),
+	} );
+};
```
Comment (youknowriad) [inline packages/blocks/src/store/actions.js:75]: if you look closely, you'll notice that basically, the logic that was in the reducers, spread across multiple reducers is now inside the action that triggered the change. I like the explicitness personally, that said, if we want to share the same logic across actions, we should be able to move this logic to the "set" function of the atoms.

Comment (youknowriad) [reply packages/blocks/src/store/actions.js:75]: In this particular store, I didn't feel the need to share behavior between actions.

File: packages/blocks/src/store/selectors.js [hunk 3]
```diff
@@ -22,203 +37,168 @@ import {
...
  *
- * @param {Object}                state     Data state.
  * @param {string}                blockName Block type name.
  * @param {WPBlockVariationScope} [scope]   Block variation scope name.
  *
  * @return {(WPBlockVariation[]|void)} Block variations.
  */
-export function getBlockVariations( state, blockName, scope ) {
-	const variations = state.blockVariations[ blockName ];
+export const getBlockVariations = ( blockName, scope ) => ( { get } ) => {
```
Comment (youknowriad) [inline packages/blocks/src/store/selectors.js:88]: While working on this selector and other similar selectors, I realized something important. Right now these selectors recompute on each mapSelectToProps but ideally they are built using `WPAtomFamily` technique to recompute only when their dependencies change.

The problem is that `WPAtomFamily` only accepts one argument for the "key" :) while selectors have an arbitrary number of arguments. So we need something like `WPAtomFamilySelector` here and if you think about it `WPAtomFamily` is just a specific case (a selector with only one argument). So I think the API would gain in clarity if we remove `WPAtomFamily` and implement a more generic ` WPAtomSelector`. I'll explore that separately.

**Comment by youknowriad** [comment]
I also intentionally didn't touch the tests to check the additions and removals of the production code itself and it seems there's a little bit less code with atomic stores. I kept the logic generally equivalent to what we had before.

**Comment by youknowriad** [comment]
the tests are failing, after some debugging, I discovered that the registry inheritance is breaking the atomic stores. I'm not sure yet how to solve that, I need some time to think about it :)

File: packages/blocks/src/store/selectors.js (hunk 3 above)
Comment (youknowriad) [reply packages/blocks/src/store/selectors.js:88]: The selector issue is being solved here https://github.com/WordPress/gutenberg/pull/27147

**Comment by youknowriad** [comment]
It is interesting to note that the "loading" time is worse. I think it's important to try to understand the reasons before proceeding.

**Comment by jsnajdr** [comment]
@youknowriad The performance problem you mention, do they boil down to selector calls being slower with `stan`? I.e., calling
```js
atomRegistry.get( getBlockVariations( blockName, scope ) );
```
is slower than analogous Redux code:
```js
getBlockVariations( store.getState(), blockName, scope );
```

In that case, looking at the code, I see several opportunities to make it faster:

**`EquivalentKeyMap` has a lot of overhead we don't need**
Given that Redux state is made of nested objects/maps, the Redux selector does just two map lookups to retrieve the value:
```js
state.get( blockName ).get( scope );
```
On the other hand, `EquivalentKeyMap` does at least six lookups, in case all args are primitive values, and much more if they are objects. There is also an `Object.keys( args ).sort()` call on every lookup.

And all this extra work is there just to support deep equality comparisons when looking up `args` values. I would argue we don't need this: Redux selectors and state trees also accept only args that can be `Map` keys, and still are able to implement great things. Deep equality has a huge performance penalty, that's why it's common in React/Redux world to do only shallow comparisons and compare even objects with `===`.

For an atom selector with N args, all we need is a set of N nested `Maps`. There's no reason why the lookup should be inherently slower than Redux.

**Redundant `Map` lookups**
The following pattern is suprisingly frequent in the code, even in `equivalent-key-map` which is supposed to be super-optimized:
```js
if ( map.get( key ) ) {
  map.set( key, new Value() );
}
const value = map.get( key );
```
There are two `map.get`s, but you can do with only one:
```js
let value = map.get( key );
if ( ! value ) {
  value = new Value();
  map.set( value );
}
```
The `.get()` calls are probably cheap anyway, but if we're measuring microseconds in an inner loop, optimizing them could have an impact.

**Comment by youknowriad** [comment]
@jsnajdr I agree that is the bottleneck and there are definitely several opportunities there but it doesn't come down just to optimizing  equivalent-key-map:

 - I did a check by removing `equivalent-key-map` and adding a `getKey` function for each selector to compute a scalar key based on arguments and use a regular map. The result was way quicker but still not performant enough compared to master. I believe this `getKey` API is something we could optionally offer as optimizations for certain selectors.

 -  Nesting: calling a selector `get( selector( something )` often comes down to multiple `registry.get( something )` and not just one depending on how much dependencies exist between selectors. You could argue that a guideline could be to avoid nesting and just rely on the "root" atoms but this is actually bad for performance in situations like that:
   - Have a selector to fetch a block object
   - Have multiple selectors based on that block object
 Nesting here allows us to not compute the derived selectors (multiple) at all when the block object stays the same. So basically nesting is good in most situations but has a cost on the initial unsubscribed calls.

So this is the exact reason I think we should do more tests, especially with the `core/block-editor` package to judge the performance impact properly. In theory, at least, thanks to nesting it should make rendering huge posts way faster but since we have async mode already in place and with the knowledge that calling atom selectors will always be a bit slower that calling redux ones, I'm really unsure what the final picture will look like. Unfortunately refactoring block-editor as an atom store, is not an easy task and it will take me some days (with parallel work), I'm at like 30% now.

**Comment by jsnajdr** [comment]
> I did a check by removing `equivalent-key-map` and adding a `getKey` function for each selector to compute a scalar key based on arguments and use a regular map. The result was way quicker but still not performant enough compared to master.

That still means that the `args` need to be serialized into a string on every lookup. Even for primitive args (strings or numbers) that's a slow operation that a Redux selector doesn't need to do. For objects and arrays it's even slower, because it does a deep traversal and implements a deep equality check.

Maintaining a nested tree of `Map`s should be much faster, as it doesn't convert the keys in any way. And it's also how an equivalent Redux store would be structured.

> Nesting: calling a selector `get( selector( something )` often comes down to multiple `registry.get( something )` and not just one depending on how much dependencies exist between selectors.

That's true, but it seems to me it's not very different from what the same Redux `data` selector would do. A Redux selector also does registry lookups when executing `select( 'core/blocks' )` and then does a few other map lookups when accessing a nested state property like `state.a.b.c`.

Somewhere you said that the `stan` approach is inherently slower and that it's a design tradeoff, not an implementation problem. I don't understand why this should be true. When I unroll both abstractions and look at the primitive ops that need to be done, I see the same sequence of lookups for both `stan` and Redux. You look up the selector atom state by one lookup (equivalent to `select( dataStore )`, then descend down a nested map (equivalent to chained prop access in `state`) and then retrieve a value.

I see only two extra steps: calling the selector (`get( mySelector( 1, 2, 3 ) )`) creates and intermediate object with shape `{ type: 'selector', config, args: [ 1, 2, 3 ] }`. A Redux selector doesn't need that, it's usually zero-allocation code. Second, the actual atom value is not a primitive value like in Redux state, but a `WPAtom` wrapper object.

I can see how the `stan` approach can do one or two extra map lookups and allocations, but not much more. Do you have some small counterexample that would show the extra performance issues that are inherent?

**Comment by youknowriad** [comment]
> Maintaining a nested tree of Maps should be much faster, as it doesn't convert the keys in any way. And it's also how an equivalent Redux store would be structured.

I thought internally equivalent-key-map used a nested tree of maps :) If it's not the case we should definitely use an alternative that works that way. (using weak map for objects and map for scalars).

That said, I know we do have use-cases where we want to consider equality `getEntityRecords( something, something, { queryObject } )` In this case, ideally the query object shouldn't be checked using strict equality as often times these objects are built inside the mapSelect function. That said, this is definitely an edge case that can be achieved using an optional optimization API.

> Somewhere you said that the stan approach is inherently slower and that it's a design tradeoff, not an implementation problem. 

Doing `state.items[ key ][property]` is way faster than 

```
const itemsSelector = createSelector( () => ({get}) => get( selector( items ) ) );
const itemSelector = createSelector( ( key ) => ({get}) => get( itemsSelector )[ key ] );
const itemPropertySelector = createSelector( ( key, property ) => ({get}) => get( itemSelector(key ) )[ property ] );
get( itemPropertySelector( key, property ) )
```

Each of the nested `get` calls need to go through the registry to find the selector state, add it to the dependencies, return the current value. I didn't compare 1-1 but that seems more work.

**Comment by jsnajdr** [comment]
> I thought internally equivalent-key-map used a nested tree of maps :)

It does, but a much much bigger one 🙂 Each `EquivalentKeyMap` has three inner maps for various types of keys (primitives, arrays, objects), and nested `EquivalentKeyMap`s are used inside these as values.

`map.get( [ 'a', 'b', 'c' ] )` doesn't unroll to a simple `_map.get( 'a' ).get( 'b' ).get( 'c' )`, but to a more complicated
```
_map.get( '0' ).get( 'a' ).get( '1' ).get( 'b' ).get( '2' ).get( 'c' )
```
where some of the `get`s are not native `Map` lookups, but further `EquivalentKeyMap` nested ones. The `0`, `1`, and `2` are array indices, i.e., `Object.keys( key )` values of the array key.

**Comment by jsnajdr** [comment]
> I thought internally equivalent-key-map used a nested tree of maps :)

One way to explain the algorithm is: take the key, e.g., `[ 1, 2, { a: 3 } ]` and serialize it into an array of tokens:
```js
[ beginArray, 1, 2, beginObject, 'a', 3, endObject, endArray ]
```
And then use this array as a path to descend into a deep tree, and retrieve the value from the leaf node.

Quite similar to just serializing the key to a string, but instead of a string key lookup in a flat map, there is a nested tree.

This algorithm description is also a recipe for a major rewrite of the implementation, making it much simpler 🙂

**Comment by jsnajdr** [comment]
> Doing `state.items[ key ][ property ]` is way faster than [...]

I see. I was considering only a selector that selects from the `items` array directly:
```js
const itemPropertySelector = createSelector( ( id, property ) => ( { get } ) => get( items )[ id ][ property ] );
```
without an intermediate `getItem` selector. I also now understand better why you say there is a tradeoff in the approach itself.

I did some analysis of what exactly a call like `get( itemPropertySelector( id, property ) )` does. Mainly to get some clarity for myself, hopefully it's useful for others, too:

First of all, the `itemsSelector` in your example doesn't seem to do anything. A selector that has no args can be simply a derived atom, and if a derived atom returns the value of another atom unchanged, then it's an identity mapping and can be eliminated. I'll continue to work only with a base `items` atom and two selectors:
```js
const itemSelector = createSelector( ( id ) => ( { get } ) => get( items )[ id ] );
const itemPropertySelector = createSelector( ( id, property ) => ( { get } ) => get( itemSelector( id ) )[ property ] );
```

Also, when talking about atom state in the registry, I'll be putting state of both atoms and selectors into one map. There don't need to be two separate WeakMaps for `atoms` and `selectors`. The keys are compatible and live in one map.

In the beginning, the atoms state map contains just one entry, for the `items` atom:
```
[ items ] => { value: [ 🍕, 🍔, 🍟 ] }
```

Now, what happens when I call `get( itemPropertySelector( 1, 'name' ) )`? Calling the `itemPropertySelector` function, and then trying to `get` its state, creates a new state entry:
```
[ items ] => { value: [ 🍕, 🍔, 🍟 ] }
[ itemPropertySelectorConfig ] => ID:1 => P:name => atom( item1name )
```
The key is the `itemPropertySelector`'s `config` object, and the value is a nested map, keyed by the args, and the value is a newly created (derived) atom for the `item1name` value. It's an atom creator, not an atom state.

`get`-ting the state of this `item1name` creates another state entry:
```
[ items ] => { value: [ 🍕, 🍔, 🍟 ] }
[ itemPropertySelectorConfig ] => ID:1 => P:name => atom( item1name )
[ item1name ] => { value: null }
```
and the `item1name` atom starts resolving. The resolver calls the `get( itemSelector( 1 ) )`, so we go through a similar process again, adding two state entries:
```
[ items ] => { value: [ 🍕, 🍔, 🍟 ] }
[ itemPropertySelectorConfig ] => ID:1 => P:name => atom( item1name )
[ item1name ] => { value: null }
[ itemSelectorConfig ] => ID:1 => atom( item1 )
[ item1 ] => { value: null }
```
Now, resolving the state of `item1` is easy: it looks at the state of existing, already resolved `items` atom and does an `id` property lookup on it.

The final state looks like:
```
[ items ] => { value: [ 🍕, 🍔, 🍟 ] }
[ itemPropertySelectorConfig ] => ID:1 => P:name => atom( item1name )
[ item1name ] => { value: 'pizza' }
[ itemSelectorConfig ] => ID:1 => atom( item1 )
[ item1 ] => { value: 🍕 }
```
and `get( itemPropertySelector( 1, 'name' ) )` can finally return `'pizza'`.

To perform this read, we needed to do the following operations:
- add 4 new items to the atom state map
- create two new 'atom' objects
- create three `Map` objects, with one entry each
- create two new 'atom state' objects and establish two subscriptions to their parent atoms

On a repeated call to `get( itemPropertySelector( 1, 'name' ) )`, this whole structure is already in place and we do only 4 map lookups:
- lookup selector state by its `config` object
- two lookups in nested maps (one for `id`, one for `name`) to find the value atom
- lookup atom state

That's not so awful compared to a Redux selector's two map lookups. It's constructing the atom state structure that's expensive and Redux doesn't need to do it at all.

---

What do we get in return for this extra work? The most visible thing are focused subscriptions. A selector for a specific property of a specific post provides not only the `get`-ter, but also an ability to subscribe to changes of that property only:
```js
subscribe( postPropertySelector( 1, 'title' ), () => {
  // by the way, why don't we get `value` directly as a listener param?
  setState( get( postPropertySelector( 1, 'title' ) ) );
}
```

Then we can also set the property directly through the selector:
```js
set( postPropertySelector( 1, 'title' ), 'Hello World!' );
```
This write then trickles up to the post and post list atom, updating them, and also triggering save requests to REST API.

In a sense, we're paying for this functionality upfront, when calling the `get`-ter. Redux selector returns the pure primitive value, without any overhead or wrapper objects.

There are several possible ways how we could optimize this:

**Don't pay for `subscribe` and `set` upfront** When calling `get`, do only the absolutely minimal work needed to get the value, without building the entire atom state structure. If, for example, we know that we will never be subscribing to `postPropertySelector`, can we have a special type of atom that supports only `get` and throws a `not supported` error on any other op? Or can we do the necessary initialization only on first call to `subscribe` or `set`? Could we use the [flyweight pattern](https://en.wikipedia.org/wiki/Flyweight_pattern) to optimize atom allocation? This is a half-baked idea and I don't have any specific code suggestions yet.

**Carefully optimize the atom allocation** Accept the atom state allocations described above, but optimize them. Besides the `EquivalentKeyMap` optimizations we already talked about, there's also an opportunity to optimize the `createAtom` and `createDerivedAtom` functions: right now, they use a pattern:
```js
function create() {
  let instanceProp;
  function method() {
    instanceProp = ...
  }
  return { method }
}
```
where the `method` is created anew on each instantiation. Maybe it's faster to use classes, where the methods are set once on a prototype, forever, and instances call always the same function, only with different `this`? And how fast is the closure with `instanceProp` and accessing it? I've seen some super-optimized libraries in the past where the authors cared a lot about similar things, and if the compiler is able to optimize certain constructs or not.

---

@youknowriad is there a simple-to-setup perf benchmark that I could setup locally and get relevant results out of it? How do you get authoritative perf metrics when testing various approaches?

**Comment by youknowriad** [comment]
Thanks for the details @jsnajdr that was exactly my analysis, I couldn't have explained it better.

> @youknowriad is there a simple-to-setup perf benchmark that I could setup locally and get relevant results out of it? How do you get authoritative perf metrics when testing various approaches?

I don't have granular performance metrics, I was essentially relying on the "Permormance Tests" job on this PR. It can fluctuate a little bit but in general, it gives a good idea wholistically.  Also I've did some comparison between the graphs I get from chrome performance monitoring when loading the page (and the difference when registering blocks is very noticeable there)

**Comment by jsnajdr** [comment]
> I don't have granular performance metrics, I was essentially relying on the "Permormance Tests" job on this PR

@youknowriad I tried to create a focuses micro-benchmark for Stan selectors and published the results in #27299. I don't know how relevant to real-world performance that benchmark is -- it's an experiment everything is up for discussion 🙂

**Comment by youknowriad** [comment]
I'm going to close this, unfortunately, with all the other priorities, I don't have enough time to move this forward.

**Comment by jsnajdr** [comment]
@youknowriad I wanted to ask one little thing about the `core/blocks` store: pretty much everything there is mostly read-only during the entire lifetime of the editor, is that right? You register all the block types and styles etc at the beginning, and since then, except an occasional block installed from block directory, the data doesn't change.

That would suggest that the data structure should be something that's really fast to read, that we don't need to care much if the occasional update is inefficient, and that we should avoid sending components that only read from `core/blocks` too many too broad change events that are almost never relevant.

**Comment by youknowriad** [comment]
> I wanted to ask one little thing about the core/blocks store: pretty much everything there is mostly read-only during the entire lifetime of the editor, is that right?

Yes, one dream goal at some point was to be able to reactively parse the content and allow block registration at any time. That said, it seems like the ship has sailed for that

**Comment by jsnajdr** [comment]
> Yes, one dream goal at some point was to be able to reactively parse the content and allow block registration at any time.

I thought that one major blocker for that were the block transforms (`raw`, `enter`, `prefix`) where I need all blocks to get registered first and only then I'm able to reliably match the transform rules.

**Pull request closed without merging**